exporter/uptraceexporter/exporter.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2019 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package uptraceexporter
+
+import (
+	"context"
+	"encoding/binary"
+
+	"github.com/uptrace/uptrace-go/spanexp"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/otel/label"
+	"go.uber.org/zap"
+)
+
+type traceExporter struct {
+	cfg    *Config
+	logger *zap.Logger
+	upexp  *spanexp.Exporter
+}
+
+func newTraceExporter(cfg *Config, logger *zap.Logger) (*traceExporter, error) {
+	upexp, err := spanexp.NewExporter(&spanexp.Config{
+		DSN: cfg.DSN,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	exporter := &traceExporter{
+		cfg:    cfg,
+		logger: logger,
+		upexp:  upexp,
+	}
+
+	return exporter, nil
+}
+
+// pushTraceData is the method called when trace data is available.
+func (e *traceExporter) pushTraceData(ctx context.Context, traces pdata.Traces) (int, error) {
+	outSpans := make([]spanexp.Span, 0, traces.SpanCount())
+
+	rsSpans := traces.ResourceSpans()
+	for i := 0; i < rsSpans.Len(); i++ {
+		rsSpan := rsSpans.At(i)
+		resource := keyValueSlice(rsSpan.Resource().Attributes())
+
+		ils := rsSpan.InstrumentationLibrarySpans()
+		for j := 0; j < ils.Len(); j++ {
+			ilsSpan := ils.At(j)
+			lib := ilsSpan.InstrumentationLibrary()
+
+			spans := ilsSpan.Spans()
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				outSpans = append(outSpans, spanexp.Span{})
```

Comment:
AFAIK that this removes lots of allocations. This way I have a slice of structs that are allocated in batches (https://www.darkcoding.net/software/go-how-slices-grow/).

>Why aren't you first building the span (out) and then appending to the list of spans?

Then I need to allocate a separate struct and then copy it to a slice. That is an allocation for each struct. Or using a value instead of a pointer and use more CPU for copying.

Reply:
Didn't know about that, thanks!